dev/kube/kube-ready-state-check.sh

```diff
@@ -126,17 +148,6 @@ if having_category kube ; then
     status "A storage class should exist in K8s"
 fi
 
-# privileged pods are enabled in K8s
```

Comment:
You also removed this check for the privileged container (`apiserver --allow-privileged`) - that needs to be put back in, I think? Not sure what to do with the kubelet privileged check as it only applies to kube versions < 1.15

Reply:
I'm not sure how robust the checks using `pgrep -ax 'hyperkube|apiserver'` really are.

Some failure modes:

* the process is not called `apiserver`, but e.g. `kube-apiserver`
* the flag is listed, but set to `false` via e.g. `--allow-privileged=false`

The first is true on CaaSP 4. The second isn't, but you would not catch it if it was:

```
vagrant@caasp4-master-1:~> pgrep -ax kube-apiserver
7463 kube-apiserver --advertise-address=192.168.121.120 --allow-privileged=true --authorization-mode=Node,RBAC --client-ca-file=/etc/kubernetes/pki/ca.crt --enable-admission-plugins=NamespaceLifecycle,LimitRanger,ServiceAccount,TaintNodesByCondition,Priority,DefaultTolerationSeconds,DefaultStorageClass,PersistentVolumeClaimResize,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota,StorageObjectInUseProtection,RuntimeClass,NodeRestriction,PodSecurityPolicy --enable-bootstrap-token-auth=true --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt --etcd-certfile=/etc/kubernetes/pki/apiserver-etcd-client.crt --etcd-keyfile=/etc/kubernetes/pki/apiserver-etcd-client.key --etcd-servers=https://127.0.0.1:2379 --insecure-port=0 --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --oidc-ca-file=/etc/kubernetes/pki/ca.crt --oidc-client-id=oidc --oidc-groups-claim=groups --oidc-issuer-url=https://caasp4-lb-1:32000 --oidc-username-claim=email --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key --requestheader-allowed-names=front-proxy-client --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-group-headers=X-Remote-Group --requestheader-username-headers=X-Remote-User --secure-port=6443 --service-account-key-file=/etc/kubernetes/pki/sa.pub --service-cluster-ip-range=10.96.0.0/12 --tls-cert-file=/etc/kubernetes/pki/apiserver.crt --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
```

On CaaSP `kube-apiserver`, `kube-scheduler` etc. all run inside pods started by `kubelet`:

```
vagrant@caasp4-master-1:~> sudo grep -r allow-privileged /etc/kubernetes/manifests
/etc/kubernetes/manifests/kube-apiserver.yaml:    - --allow-privileged=true
```

I assume the objection did not extend to the `kubelet` check, as it's `--allow-privileged` option was removed in kube 1.15.

Reply:
> * the process is not called `apiserver`, but e.g. `kube-apiserver`

This could be fixed by `pgrep -ax 'hyperkube|(kube-)?apiserver'`, but it remains a heuristic, which means it sometimes will still not do the right thing.

Reply:
I agree with your first point - changes pushed already.
However, the second point might be bit of a stretch as we cannot account for all the possible configurations on different platforms.

Reply:
> the second point might be bit of a stretch as we cannot account for all the possible configurations on different platforms.

I don't know what the second point is that you refer to here, but the whole point of this script is checking any arbitrary configuration for compatibility with kubecf. If you don't mind detecting the values correctly, why bother checking for them at all.

Here is a simple test that will treat `--allow-privileged=false` correctly:

```
# echo '--allow-privileged' | perl -ne 'exit(!/--allow-privileged(?!(=| +)false)/)'; echo $?
0
# echo '--allow-privileged --foo' | perl -ne 'exit(!/--allow-privileged(?!(=| +)false)/)'; echo $?
0
# echo '--allow-privileged=true' | perl -ne 'exit(!/--allow-privileged(?!(=| +)false)/)'; echo $?
0
# echo '--allow-privileged=false' | perl -ne 'exit(!/--allow-privileged(?!(=| +)false)/)'; echo $?
1
# echo '--allow-privileged  false' | perl -ne 'exit(!/--allow-privileged(?!(=| +)false)/)'; echo $?
1
```

It uses a negative look-ahead assertion to make sure `--allow-privileged` is not followed by `=` or one or more spaces, followed by the string `false`.

Reply:
updated check.